BUG: Switched shapes in ValueError msg in DataFrame construct (#20742)

meiermark opened the pull request:
- [x] closes #20742
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

meiermark committed:
BUG: Switched shapes in ValueError msg in DataFrame construct (#20742)

gfyoung commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1852,6 +1852,7 @@ Other
 ^^^^^
 
 - Bug where C variables were declared with external linkage causing import errors if certain other C libraries were imported before Pandas. (:issue:`24113`)
+- Switched shape in ``ValueError`` message when constructiong a :class:`DataFrame` with parameters ``columns`` and ``index`` not matching the shape of the input data. (:issue:`20742`)
```
Let's be explicit as to how the dimensions are reported (i.e. their order) after the switch.

meiermark committed:
DOC: improved docu of the value error msg changes

TST: adjusted panel test for change in value error msg in block manager
TST: adjusted dataframe init test from json for change in value error
     msg in block manager

meiermark committed:
BUG: printed a list instead of a tuple in the ValueError msg

meiermark committed:
CLN: removed unnecessary list comprehension

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -1673,8 +1673,10 @@ def create_block_manager_from_arrays(arrays, names, axes):
     """ raise a helpful message about our construction """
-    passed = tuple(map(int, [tot_items] + list(block_shape)))
-    implied = tuple(map(int, [len(ax) for ax in axes]))
+    passed = tuple(map(int, list(block_shape) + [tot_items]))
```
just do a reverse if ndim == 2 here

meiermark replied:
Probably not a good idea for panels with 3 dimensions (items, major, minor)
with reversed for ndim == 2: (items, major, minor)
current: (major, minor, items)

The current version is the intended behavior, right?

jreback replied:
I don't care about Panels, they are being removed shortly.

jreback commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -386,25 +386,37 @@ def test_constructor_error_msgs(self):
                       columns=['foo', 'bar', 'baz'],
                       index=pd.date_range('2000-01-01', periods=3))
 
+        # see issue #20742
```
you don't need the issue number

meiermark replied:
okay.

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1143,23 +1143,11 @@ def test_from_dict_mixed_orient(self):
         with pytest.raises(ValueError, match=msg):
             Panel(np.random.randn(3, 4, 5),
-                  lrange(4), lrange(5), lrange(5))
-
```
leave these tests here

meiermark replied:
okay. Is there additional value in those tests?

meiermark committed:
DOC: Improved whatsnew message

meiermark commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1852,6 +1852,9 @@ Other
 
 - Bug where C variables were declared with external linkage causing import errors if certain other C libraries were imported before Pandas. (:issue:`24113`)
+- Changed ``ValueError`` message when constructiong a :class:`DataFrame` with parameters ``columns`` and ``index`` not matching the shape of the input data. New message: "Shape of passed values is (``index``, ``columns``), indices imply (``index``, ``columns``)" (:issue:`20742`)
+- Changed ``ValueError`` message when constructiong a :class:`Panel` with parameters ``items``, ``major index`` and ``minor index`` not matching the shape of the input data. New message: "Shape of passed values is (``major``, ``minor``, ``items``), indices imply (``major``, ``minor``, ``items``)" (related to :issue:`20742`)
```
I improved the message. It's now shorter. However, I am not sure if it is short enough or which parts you could still remove.

meiermark committed:
CLN: readded deleted tests

CLN: removed issue number above new tests

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1143,20 +1143,20 @@ def test_from_dict_mixed_orient(self):
     def test_constructor_error_msgs(self):
-        msg = (r"Shape of passed values is \(3, 4, 5\), "
-               r"indices imply \(4, 5, 5\)")
+        msg = (r"Shape of passed values is \(4, 5, 3\), "
```
leave panel entirely.

meiermark committed:
Bug: No changes for panel

TomAugspurger reviewed (commented):
Need to fix a merge conflict.

meiermark committed:
DOC: improved whatsnew message for reversed shapes

meiermark committed:
Merge branch 'master' into dataframe_creation_error_reversed_shape

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -1674,7 +1674,13 @@ def create_block_manager_from_arrays(arrays, names, axes):
     """ raise a helpful message about our construction """
     passed = tuple(map(int, [tot_items] + list(block_shape)))
-    implied = tuple(map(int, [len(ax) for ax in axes]))
+    if len(passed) == 2:
```
can you add a comment on the reversing. (e.g. this is a user facing error message and this is a DataFrame construction)

you can do len(passed) <= 2

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1816,6 +1816,7 @@ Reshaping
 - Bug in :func:`DataFrame.unstack` where a ``ValueError`` was raised when unstacking timezone aware values (:issue:`18338`)
 - Bug in :func:`DataFrame.stack` where timezone aware values were converted to timezone naive values (:issue:`19420`)
 - Bug in :func:`merge_asof` where a ``TypeError`` was raised when ``by_col`` were timezone aware values (:issue:`21184`)
+- Reversed shapes in ValueError message when data shape does not match index shape in construction. (:issue:`20742`)
```
say Incorrect shape in DataFrame construction error message.

meiermark committed:
DOC: improved comments and whatsnew message for df construction error

jreback reviewed (changes requested):
whatsnew correction. ping on green.

meiermark committed:
DOC: improved whatsnew message

TomAugspurger commented:
Thanks @meiermark.
